### PR TITLE
Limit the number of subscriptions for a websocket connection

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

Too many subscriptions `klay_subscription` can delay block synchronization or cause Out-of-Memory of Klaytn nodes. 
For ENs used in enterprise services, Klaytn nodes need to manage the subscription number on the server-side. 
This PR limits the maximum subscription number for a server connection to 10.

If a user requests more than 10 active subscriptions, he/she will get the following error.
```
{
  "jsonrpc": "2.0",
  "id": 0,
  "error": {
    "code": -32000,
    "message": "Maximum 10 subscriptions are allowed for a websocket connection"
  }
```

~TODO: consider obtaining `maxSubscriptionPerConn` from a flag value.~ 
=> New API `admin_setMaxSubscriptionPerConn` is introduced and tested on my local environment. 
\+ New flag `wsmaxsubscriptionperconn` is added. 

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ehnuje — DISMISSED** [review]

File: networks/rpc/server.go [hunk 1]
```diff
@@ -331,6 +337,7 @@ func (s *Server) handle(ctx context.Context, codec ServerCodec, req *serverReque
 				return codec.CreateErrorResponse(&req.id, &callbackError{err.Error()}), nil
 			}
 
+			*subCnt--
```
Comment (winnie-byun) [inline networks/rpc/server.go:340]: Shouldn't this be ++?
```suggestion
			*subCnt++
```

**Comment by winnie-byun** [comment]
Has this code been tested?

File: networks/rpc/server.go (hunk 1 above)
Comment (ehnuje) [reply networks/rpc/server.go:340]: @winnie-byun Please check line 326, it is unsubscribe branch.

File: networks/rpc/server.go [hunk 2]
```diff
@@ -49,6 +49,9 @@ const (
 
 	// concurrencyLimit is a limit for the number of concurrency connection for RPC servers
 	concurrencyLimit = 3000
+
+	// maxSubscriptionPerConn is a maximum number of subscription for a server connection
+	maxSubscriptionPerConn = 10
```
Comment (winnie-byun) [inline networks/rpc/server.go:54]: I think this value should be parameterized.
This limits the number of subscription and a node operator might wants to subscribe more than 10 times.

File: networks/rpc/server.go [hunk 3]
```diff
@@ -350,6 +363,7 @@ func (s *Server) handle(ctx context.Context, codec ServerCodec, req *serverReque
 			notifier, _ := NotifierFromContext(ctx)
 			notifier.activate(subid, req.svcname)
 		}
+		*subCnt++
```
Comment (ehnuje) [inline networks/rpc/server.go:366]: Isn't atomic operation needed here?

Comment (winnie-byun) [reply networks/rpc/server.go:366]: Doesn't a lock needed for `subCnt`? it is accessed from several go routines.

Comment (KimKyungup) [reply networks/rpc/server.go:366]: I also agree with Melvin's comment.

Comment (aidan-kwon) [reply networks/rpc/server.go]: You are right! Thanks! I changed it to atomic operation.

**Review by ehnuje — APPROVED** [review]

**Comment by aidan-kwon** [comment]
> Has this code been tested?

It was tested on my local machine with some scenarios I showed you on another channel.

**Review by jeongkyun-oh — DISMISSED** [review]

**Review by winnie-byun — DISMISSED** [review]

**Review by yoomee1313 — APPROVED** [review]
Thank you :)

**Review by KimKyungup — COMMENTED** [review]
LGTM.
After hotfix, if we parameterize this `maxSubscriptionPerConn`, I think we can merge this.
Thanks a lot.

**Comment by aidan-kwon** [comment]
Reviewers, `maxSubscriptionPerConn` is changed from 10 to 5. If you have other opinions, please let me know.

**Review by jeongkyun-oh — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Review by yoomee1313 — APPROVED** [review]

**Review by winnie-byun — DISMISSED** [review]

**Comment by aidan-kwon** [comment]
Reviewers, I introduced `admin_setMaxSubscriptionPerConn` to configure `MaxSubscriptionPerConn` easily. PTAL

**Review by ehnuje — APPROVED** [review]

File: networks/rpc/server.go [hunk 4]
```diff
@@ -51,8 +51,13 @@ const (
 	concurrencyLimit = 3000
 )
 
-// pendingRequestCount is a total number of concurrent RPC method calls
-var pendingRequestCount int64 = 0
+var (
+	// pendingRequestCount is a total number of concurrent RPC method calls
+	pendingRequestCount int64 = 0
+
```
Comment (KimKyungup) [inline networks/rpc/server.go:57]:
```suggestion
var (
```

**Review by KimKyungup — COMMENTED** [review]
LGTM. 
Please write a PR for klaytn-docs.

File: networks/rpc/server.go (hunk 4 above)
Comment (aidan-kwon) [reply networks/rpc/server.go:57]: @KimKyungup Ths suggestion removes `pendingRequestCount `. Do you want to declare `pendingRequestCount` `MaxSubscriptionPerConn` seperatly?

**Comment by KimKyungup** [comment]
@aidan-kwon How about adding an option for this?

When we set up the KES service node, it is hard to set this value via API everytime.

**Comment by aidan-kwon** [comment]
> @aidan-kwon How about adding an option for this?
> 
> When we set up the KES service node, it is hard to set this value via API everytime.

@KimKyungup I am not sure it is adequate for the node flag, but I agree with your point. Your suggestion is applied.

**Review by winnie-byun — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

File: networks/rpc/server.go (hunk 4 above)
Comment (KimKyungup) [reply networks/rpc/server.go:57]: It's my mistake. Please forget it. :)

**Review by KimKyungup — APPROVED** [review]

**Pull request merged**